File: Julia/life.jl
# A Julia implementation of Conway's game of life

# A function to display each line
function showline(x)
    for i in x
        if i == 1
            print("# ")
        else
            print(". ")
        end
    end
    print("\n")
end

# A function to count the live neighbours of a cell
function countneighbours((x, y), grid)
    n = 0
    for i in -1:1
        for j in -1:1
            if i == 0 && j == 0 # Ignore central cell
                continue
            elseif x + i < 1 || y + j < 1 # goes out of bounds
                continue
            elseif x + i > size(grid, 1) || y + j > size(grid, 2)
                continue
            elseif grid[x+i, y+j] == 1 # valid position, now check if it is live
                n += 1
            end
        end
    end
    return n
end

# A recursive function that runs the game of life. Args of grid, steps
function life(grid, steps)

    if steps <= 0  # Break if it reaches the end
        return
    end

    newgrid = Array{Int}(undef, size(grid)) # Create new grid to avoid conflict

    for x in 1:size(grid, 1) # Apply rules
        for y in 1:size(grid, 2)
            liveneighbours = countneighbours((x, y), grid)  # Find live neighbours
            if grid[x, y] == 1 && (liveneighbours == 2 || liveneighbours == 3) # Cell survives if it has 2 or 3 neighbours
                newgrid[x, y] = 1
            elseif grid[x, y] == 0 && liveneighbours == 3 # Cell is born if it has 3 live neighbours
                newgrid[x, y] = 1
            else # Otherwise the cell is dead. Combines several rules into this one
                newgrid[x, y] = 0
            end
        end
    end

    for r in eachrow(newgrid) # Print grid
        showline(r)
    end
    println("\n")

    life(newgrid, steps-1) # return ecursive call
end

# Pre-defined patterns

# Oscillators
blinker = [0 0 0 0 0;
           0 0 0 0 0;
           0 1 1 1 0;
           0 0 0 0 0;
           0 0 0 0 0]

toad = [0 0 0 0 0 0;
        0 0 0 0 0 0;
        0 0 1 1 1 0;
        0 1 1 1 0 0;
        0 0 0 0 0 0;
        0 0 0 0 0 0]

beacon = [0 0 0 0 0 0;
          0 1 1 0 0 0;
          0 1 1 0 0 0;
          0 0 0 1 1 0;
          0 0 0 1 1 0;
          0 0 0 0 0 0]

pulsar = [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0;
          0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0;
          0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0;
          0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0;
          0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0;
          0 0 0 0 0 0 1 1 0 0 0 1 1 0 0 0 0 0 0 0;
          0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0;
          0 0 1 1 1 0 0 1 1 0 1 1 0 0 1 1 1 0 0 0;
          0 0 0 0 1 0 1 0 1 0 1 0 1 0 1 0 0 0 0 0;
          0 0 0 0 0 0 1 1 0 0 0 1 1 0 0 0 0 0 0 0;
          0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0;
          0 0 0 0 0 0 1 1 0 0 0 1 1 0 0 0 0 0 0 0;
          0 0 0 0 1 0 1 0 1 0 1 0 1 0 1 0 0 0 0 0;
          0 0 1 1 1 0 0 1 1 0 1 1 0 0 1 1 1 0 0 0;
          0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0;
          0 0 0 0 0 0 1 1 0 0 0 1 1 0 0 0 0 0 0 0;
          0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0;
          0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0;
          0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0;
          0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0;]

File: Java/src/life.java
// A Java implementation of Conway's game of life
public class life {
	public static void main(String[] args) {  // Args should be grid, 
		runLife(blinker, 3);
	}
	
	// Run the game
	public static void runLife(int[][] grid, int steps) {
		while(steps > 0) {
			int[][] newgrid = new int[grid.length][grid[0].length];
			for(int i = 0; i<grid.length; i++) {
				for(int j = 0; j<grid[1].length; j++) {
					if ((neighbours(i, j, grid) == 2 || neighbours(i, j, grid) == 3) && grid[i][j] == 1) {
						newgrid[i][j] = 1; // Cell survives 
					}
					else if (neighbours(i, j, grid) == 3 && grid[i][j] == 0) {
						newgrid[i][j] = 1; // Cell is born
					}
					else {
						newgrid[i][j] = 0; // Else the cell dies
					}
				}
			}
			grid = newgrid; // Set the current grid as newgrid for the next iteration
			display(grid);
			steps--;
		}
		
	}
	
	// Display a grid
	static void display(int[][] grid) {
		for(int i = 0; i<grid.length; i++) {
			for(int j = 0; j<grid[1].length; j++) {
				if(grid[i][j] == 1) {
					System.out.print("# ");
				}
				else { 
					System.out.print(". ");
				}
			}
			System.out.print('\n');
		}
		System.out.print('\n');
	}
	
	// Finds the number of neighbours of a cell
	static int neighbours(int x, int y, int[][] grid) {
		int neighbours = 0;
		for(int i = -1; i<=1; i++) {
			for(int j = -1; j<=1; j++) {
				if ((i == 0) && (j == 0)) {  // Ignore centre
					continue;
				}
				else if ((x + i < 0) || (y + j < 0)) { // goes out of bounds
	                continue;
				}
	            else if ((x + i >= grid.length) ||( y + j >= grid[0].length)) {
	            	continue;
	            }
	            else if ((grid[x+i][y+j] == 1)) {// valid position, now check if it is live
	                neighbours++;
	            }
			}
		}
		return neighbours;
	}

	// Example patterns 
	static int[][] blinker = new int[][] {
		{0, 0, 0, 0, 0},
		{0, 0, 1, 0, 0},
		{0, 0, 1, 0, 0},
		{0, 0, 1, 0, 0},
		{0, 0, 0, 0, 0}
	};
}
